File: GCJ/2019/qualification/C.java
import java.util.*;

public class C {
    public static boolean isPrime(long x) {
        for (long i = 3; i < x / 2 + 1; i += 2) {
            if (x % i == 0) {
                return false;
            }
        }
        return true;
    }

    static List<Long> PRIMES = new ArrayList<>();

    static {
        PRIMES.add(2L);
        for (long i = 3; i < 10000; i += 2) {
            if (isPrime(i)) {

                PRIMES.add(i);
            }
        }
    }

    public static String solve(List<Long> nums) {
        List<Long> primes = new ArrayList<>();
        for (long prime : PRIMES) {
            if (nums.get(0) % prime == 0) {
                primes.add(prime);
                primes.add(nums.get(0) / prime);
                break;
            }
        }
        for (int i = 1; i < nums.size(); i++) {
            long lastPrime = primes.get(primes.size() - 1);
            primes.add(nums.get(i) / lastPrime);
        }
        TreeSet<Long> sortedPrimes = new TreeSet<>();
        sortedPrimes.addAll(primes);
        Map<Long, Character> primeToChar = new HashMap<>();

        int i = 0;
        for (long prime : sortedPrimes) {
           primeToChar.put(prime, new Character((char) ('A' + i++)));
        }

        StringBuilder builder = new StringBuilder();

        for (long prime : primes) {
            builder.append(primeToChar.get(prime));
        }
        return builder.toString();

    }

    public static void main(String [] args) {
        Scanner scanner = new Scanner(System.in);
        int T = scanner.nextInt();
        for (int case_num = 1; case_num <= T; ++case_num) {
           long N = scanner.nextLong();
           int L = scanner.nextInt();
            List<Long> nums = new ArrayList<>(L);
           for (int i = 0; i < L; i++) {
              nums.add(scanner.nextLong());
           }

           System.out.println(String.format("Case #%d: %s", case_num, solve(nums)));
        }
    }
}

File: GCJ/2019/qualification/C.py
#!/usr/bin/env python3
import sys


def is_prime(x):
    for i in range(2, x // 2 + 1, 2):
        if x % i == 0:
            return False
    return True


primes = [2] + list(filter(lambda x: is_prime(x), range(3, 20000, 2)))


def solve(nums):
    # print(nums)
    first_primes = list(factor(nums[0]))
    # print(first_primes)

    for num in nums[1:]:
        last_prime = first_primes[-1]
        first_primes.append(num // last_prime)

    prime_to_char = {}
    sorted_primes = sorted(set(first_primes))
    # print(sorted_primes)
    for i, prime in enumerate(sorted_primes):
        prime_to_char[prime] = chr(ord('A') + i)
    # print(prime_to_char)

    return "".join(map(lambda x: prime_to_char[x], first_primes))


def factor(x):
    for prime in primes:
        if x % prime == 0:
            return prime, x // prime
    return x


data = sys.stdin.readlines()
T = int(data[0])

for case_num in range(0, T):
    N, L = list(map(lambda x: int(x), data[case_num * 2 + 1].split(" ")))
    solution = solve(list(map(lambda x: int(x), data[case_num * 2 + 2].split(" ")[0:L])))
    print("Case #%d: %s" % (case_num + 1, solution))
